src/tracer/propagation/cloud_trace_propagator.cpp

```diff
@@ -0,0 +1,213 @@
+//--------------------------------------------------------------------------------------------------
+template <class KeyCompare>
+opentracing::expected<bool> CloudTracePropagator::ExtractSpanContextImpl(
+    const opentracing::TextMapReader& carrier, TraceContext& trace_context, const KeyCompare& key_compare,
+    BaggageProtobufMap& baggage) const {
+      bool parent_header_found = false;
+      auto result =
+          carrier.ForeachKey([&](opentracing::string_view key,
+                                 opentracing::string_view
+                                     value) noexcept->opentracing::expected<void> {
+            if (key_compare(key, PropagationSingleKey)) {
+              auto was_successful = this->ParseCloudTrace(value, trace_context);
+              if (!was_successful) {
+                return opentracing::make_unexpected(was_successful.error());
+              }
+              parent_header_found = true;
+            } else if (key.length() > PrefixBaggage.size() &&
+                       key_compare(opentracing::string_view{key.data(),
+                                                            PrefixBaggage.size()},
+                                   PrefixBaggage)) {
+              baggage.insert(BaggageProtobufMap::value_type(
+                  ToLower(
+                      opentracing::string_view{key.data() + PrefixBaggage.size(),
+                                               key.size() - PrefixBaggage.size()}),
+                  value));
+            }
+            return {};
+          });
+      if (!result) {
+        return opentracing::make_unexpected(result.error());
+      }
+      return parent_header_found;
+}
+
+opentracing::expected<void> CloudTracePropagator::ParseCloudTrace(
+    opentracing::string_view s, TraceContext& trace_context) const noexcept {
+  if (s.size() < Num128BitHexDigits) {
+    return opentracing::make_unexpected(
+        std::make_error_code(std::errc::invalid_argument));
+  }
+  size_t offset = 0;
+
+  // default sampled to on (this comes from the ;o=1 part of
+  // x-cloud-trace-context; if it is not set we will default to sampling
+  // this request)
+  trace_context.trace_flags = SetTraceFlag<SampledFlagMask>(trace_context.trace_flags, true);
+
+  // trace-id
+  auto error_maybe = NormalizedHexToUint128(
+      opentracing::string_view{s.data() + offset, Num128BitHexDigits},
+      trace_context.trace_id_high, trace_context.trace_id_low);
+  if (!error_maybe) {
+    return error_maybe;
+  }
+
+  offset += Num128BitHexDigits;
+  if (s.size() < Num128BitHexDigits + 2) {
+    // only a short form trace ID has been given (not a "trace id/span id")
+    return {};
+  }
+
+  if (s[offset] != '/') {
+    return opentracing::make_unexpected(
+        std::make_error_code(std::errc::invalid_argument));
+  }
+  ++offset;
+
+  char parent_id [20];
+  int span_id_length;
+  for(span_id_length = 0; span_id_length<20; span_id_length++) {
+    if(s[offset]==';' || s[offset]=='\0') {
```

Comment:
s is a string_view which may or may not use a null terminator.

This will have to change to safely handle the case where s isn't non-null terminated.

Reply:
https://github.com/lightstep/lightstep-tracer-cpp/pull/248/commits/0c0f66a84e3f2ecdcb1303ca46831d2b420b8be0

I added a condition to the for loop to make sure we don't go past the end of `s.length()`

I also fixed a problem with a missing null terminator in this commit discovered by adding a new test case with a short span ID of `123` (with `x-cloud-trace-context` the span ID is a 64bit unsigned long so can be 1-20 characters as a string)

I did try to pass this as an unterminated char array from the test but the `TextMapCarrier` or something in there was trying to add the `\0` terminator to the end of it.

Reply:
Can you change the check to
```
if(!std::isdigit(s[offset])) {
   break;
}
```

Reply:
Thanks, done